File: board.py
import pygame
from constants import *
import random

class Tetromino:
    def __init__(self, x, y, shape):
        self.x = x
        self.y = y
        self.shape = shape
        self.color = SHAPE_COLORS[SHAPES.index(shape)]
        self.rotation = 0

    def image(self):
        return self.shape[self.rotation % len(self.shape)]

    def rotate(self):
        self.rotation = (self.rotation + 1) % len(self.shape)

class Board:
    def __init__(self):
        self.locked_positions = {}
        self.grid = self.create_grid(self.locked_positions)
        self.score = 0

    def create_grid(self, locked_positions={}):
        grid = [[(0,0,0) for _ in range(10)] for _ in range(20)]
        for y in range(len(grid)):
            for x in range(len(grid[y])):
                if (x,y) in locked_positions:
                    c = locked_positions[(x,y)]
                    grid[y][x] = c
        return grid

    def convert_shape_format(self, shape):
        positions = []
        format = shape.image()

        for i, line in enumerate(format):
            row = list(line)
            for j, column in enumerate(row):
                if column == '0':
                    positions.append((shape.x + j, shape.y + i))

        for i, pos in enumerate(positions):
            positions[i] = (pos[0] - 2, pos[1] - 4)
        return positions

    def valid_space(self, shape):
        accepted_positions = [[(j, i) for j in range(10) if self.grid[i][j] == (0,0,0)] for i in range(20)]
        accepted_positions = [j for sub in accepted_positions for j in sub]
        formatted = self.convert_shape_format(shape)

        for pos in formatted:
            if pos not in accepted_positions:
                if pos[1] > -1:
                    return False
        return True

    def check_lost(self, positions):
        for pos in positions:
            x, y = pos
            if y < 1:
                return True
        return False

    def clear_rows(self, grid, locked):
        inc = 0
        for i in range(len(grid)-1,-1,-1):
            row = grid[i]
            if (0,0,0) not in row:
                inc += 1
                ind = i
                for j in range(len(row)):
                    try:
                        del locked[(j, i)]
                    except KeyError:
                        continue
        if inc > 0:
            for key in sorted(list(locked), key=lambda x: x[1])[::-1]:
                x, y = key
                if y < ind:
                    newKey = (x, y + inc)
                    locked[newKey] = locked.pop(key)
        self.score += inc * 10

File: Board.cpp
#include "Board.hpp"
#include "Constants.hpp"
#include <algorithm>

Board::Board() : score(0) {
    this->create_grid();
}

void Board::create_grid() {
    this->grid.assign(20, std::vector<Color>(10, {0.0, 0.0, 0.0}));
    for (const auto& pos_color_pair : this->locked_positions) {
        const auto& pos = pos_color_pair.first;
        const auto& color = pos_color_pair.second;
        if (pos.second >= 0 && pos.second < 20 && pos.first >= 0 && pos.first < 10) {
            this->grid[pos.second][pos.first] = color;
        }
    }
}

std::vector<std::pair<int, int>> Board::convert_shape_format(const Tetromino& shape) const {
    std::vector<std::pair<int, int>> positions;
    auto shape_template = shape.get_shape();

    for (int i = 0; i < shape_template.size(); ++i) {
        for (int j = 0; j < shape_template[i].size(); ++j) {
            if (shape_template[i][j] == '0') {
                positions.push_back({shape.x + j - 2, shape.y + i - 4});
            }
        }
    }
    return positions;
}

bool Board::valid_space(const Tetromino& shape) const {
    auto formatted_shape = this->convert_shape_format(shape);
    for (const auto& pos : formatted_shape) {
        if (pos.first < 0 || pos.first >= 10 || pos.second >= 20) {
            return false;
        }
        if (pos.second >= 0 && this->grid[pos.second][pos.first].r != 0.0) {
            return false;
        }
    }
    return true;
}

bool Board::check_lost() const {
    for (const auto& pos_color_pair : this->locked_positions) {
        if (pos_color_pair.first.second < 1) {
            return true;
        }
    }
    return false;
}

void Board::clear_rows() {
    int rows_cleared = 0;
    std::vector<int> full_rows;

    // First, find all the full rows from bottom to top
    for (int i = 19; i >= 0; --i) {
        bool row_full = true;
        for (int j = 0; j < 10; ++j) {
            if (this->grid[i][j].r == 0.0) { // Check for an empty cell
                row_full = false;
                break;
            }
        }
        if (row_full) {
            full_rows.push_back(i);
            rows_cleared++;
        }
    }

    if (rows_cleared == 0) {
        return; // Nothing to do
    }

    // Update score
    this->score += rows_cleared * 10 * rows_cleared; // Bonus for multi-line clears

    // Create a new map for locked positions
    std::map<std::pair<int, int>, Color> new_locked_positions;

    // Copy over the blocks that are not in the cleared rows
    for (const auto& pair : this->locked_positions) {
        bool in_cleared_row = false;
        for (int row_y : full_rows) {
            if (pair.first.second == row_y) {
                in_cleared_row = true;
                break;
            }
        }
        if (!in_cleared_row) {
            // Count how many cleared rows are below this block
            int shift_down = 0;
            for (int row_y : full_rows) {
                if (row_y > pair.first.second) {
                    shift_down++;
                }
            }
            new_locked_positions[{pair.first.first, pair.first.second + shift_down}] = pair.second;
        }
    }

    this->locked_positions = new_locked_positions;
}

void Board::lock_piece(const Tetromino& shape) {
    auto color_vec = shape.get_color();
    Color piece_color = {color_vec[0], color_vec[1], color_vec[2]};
    for (const auto& pos : this->convert_shape_format(shape)) {
        this->locked_positions[pos] = piece_color;
    }
    this->clear_rows();
    this->create_grid();
}
